File: ios/SourceModule.m
#import <React/RCTBridgeModule.h>

@interface RCT_EXTERN_REMAP_MODULE(SourceModule, SourceModule, NSObject)

RCT_EXTERN_METHOD(initWithConfig:(NSString *)nativeId
                  drmNativeId:(NSString *)drmNativeId
                  config:(nullable id)config
                  sourceRemotePlaybackConfig:(id)remotePlayback)
RCT_EXTERN_METHOD(initWithAnalyticsConfig:(NSString *)nativeId
                  drmNativeId:(NSString *)drmNativeId
                  config:(nullable id)config
                  sourceRemotePlaybackConfig:(id)remotePlayback
                  analyticsSourceMetadata:(nullable id)analyticsSourceMetadata)
RCT_EXTERN_METHOD(destroy:(NSString *)nativeId)
RCT_EXTERN_METHOD(
    isAttachedToPlayer:(NSString *)nativeId
    resolver:(RCTPromiseResolveBlock)resolve
    rejecter:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(
    isActive:(NSString *)nativeId
    resolver:(RCTPromiseResolveBlock)resolve
    rejecter:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(
    duration:(NSString *)nativeId
    resolver:(RCTPromiseResolveBlock)resolve
    rejecter:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(
    loadingState:(NSString *)nativeId
    resolver:(RCTPromiseResolveBlock)resolve
    rejecter:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(
    getMetadata:(NSString *)nativeId
    resolver:(RCTPromiseResolveBlock)resolve
    rejecter:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(setMetadata:(NSString *)nativeId metadata:(nullable id)metadata)
RCT_EXTERN_METHOD(
    getThumbnail:(NSString *)nativeId
    time:(nonnull NSNumber *)time
    resolver:(RCTPromiseResolveBlock)resolve
    rejecter:(RCTPromiseRejectBlock)reject)

@end

File: src/modules/SourceModule.ts
import { NativeModule, requireNativeModule } from 'expo-modules-core';
import { LoadingState, SourceRemoteControlConfig } from '../source';
import { Thumbnail } from '../thumbnail';

export type SourceModuleEvents = Record<string, any>;

declare class SourceModule extends NativeModule<SourceModuleEvents> {
  /**
   * Checks if the source is attached to a player.
   */
  isAttachedToPlayer(nativeId: string): Promise<boolean | null>;

  /**
   * Checks if the source is currently active.
   */
  isActive(nativeId: string): Promise<boolean | null>;

  /**
   * Initializes a source with the given nativeId, optional DRM nativeId, configuration object,
   * and remote control object.
   */
  initializeWithConfig(
    nativeId: string,
    drmNativeId?: string,
    config?: Record<string, any>,
    remoteControl?: SourceRemoteControlConfig
  ): Promise<void>;

  /**
   * Initializes a source with the given nativeId, optional DRM nativeId, configuration object,
   * remote control object, and analytics source metadata.
   */
  initializeWithAnalyticsConfig(
    nativeId: string,
    drmNativeId?: string,
    config?: Record<string, any>,
    remoteControl?: SourceRemoteControlConfig,
    analyticsSourceMetadata?: Record<string, any>
  ): Promise<void>;

  /**
   * Destroys the source with the given nativeId.
   */
  destroy(nativeId: string): Promise<void>;

  /**
   * Returns the native DRM config reference of the source with the given nativeId.
   */
  duration(nativeId: string): Promise<number | null>;
  /**
   * Returns the current loading state of the source with the given nativeId.
   */
  loadingState(nativeId: string): Promise<LoadingState | null>;

  /**
   * Returns the metadata of the source with the given nativeId.
   */
  getMetadata(nativeId: string): Promise<Record<string, any> | null>;

  /**
   * Sets the metadata of the source with the given nativeId.
   */
  setMetadata(
    nativeId: string,
    metadata: Record<string, any> | null
  ): Promise<void>;

  /**
   * Returns a thumbnail for the specified playback time.
   */
  getThumbnail(nativeId: string, time: number): Promise<Thumbnail | null>;
}

export default requireNativeModule<SourceModule>('SourceModule');
